testing/upload_validation.ts

```diff
@@ -933,138 +942,125 @@ const genericObjectSchema: z.ZodTypeAny = z.lazy(() =>
-            });
-            return;
-          }
-
-          const descriptionPropertySchema = schema.properties[schema.descriptionProperty];
-          if (
-            descriptionPropertySchema.type !== ValueType.String &&
-            (descriptionPropertySchema.type !== ValueType.Array ||
-              descriptionPropertySchema.items.type !== ValueType.String)
-          ) {
-            context.addIssue({
-              code: z.ZodIssueCode.custom,
-              path: ['descriptionProperty'],
-              message: `The "descriptionProperty" field name "${schema.descriptionProperty}" must refer to a "ValueType.String" property or array of strings.`,
-            });
-            return;
-          }
-        }
+      const validateSnippetProperty = () => {
+        return validateProperty(
+          'snippetProperty',
+          snippetPropertySchema =>
+            snippetPropertySchema.type === ValueType.String ||
+            (snippetPropertySchema.type === ValueType.Array && snippetPropertySchema.items.type === ValueType.String),
+          `must refer to a "ValueType.String" property or array of strings.`,
+        );
       };
       const validateLinkProperty = () => {
-        if (schema.linkProperty) {
-          if (!(schema.linkProperty in schema.properties)) {
+        return validateProperty(
+          'linkProperty',
+          linkPropertySchema =>
+            linkPropertySchema.type === ValueType.String && linkPropertySchema.codaType === ValueHintType.Url,
+          `must refer to a "ValueType.String" property with a "ValueHintType.Url" "codaType".`,
+        );
+      };
+
+      validateTitleProperty();
+      validateLinkProperty();
+      validateImageProperty();
+      validateSnippetProperty();
+      (schema.subtitleProperties || []).forEach((f, i) => {
+        if (typeof f === 'string') {
+          if (!(f in schema.properties)) {
             context.addIssue({
               code: z.ZodIssueCode.custom,
-              path: ['linkProperty'],
-              message: `The "linkProperty" field name "${schema.linkProperty}" does not exist in the "properties" object.`,
+              path: ['subtitleProperties', i],
+              message: `The "subtitleProperties" field name "${f}" does not exist in the "properties" object.`,
             });
             return;
           }
 
-          const linkPropertySchema = schema.properties[schema.linkProperty];
-          if (linkPropertySchema.type !== ValueType.String || linkPropertySchema.codaType !== ValueHintType.Url) {
+          const subtitlePropertySchema = schema.properties[f];
+          if (
+            'codaType' in subtitlePropertySchema &&
+            subtitlePropertySchema.codaType &&
+            [
+              ValueHintType.ImageAttachment,
+              ValueHintType.Attachment,
+              ValueHintType.ImageReference,
+              ValueHintType.Embed,
+              ValueHintType.Scale,
+              ValueHintType.Scale,
```

Comment:
These got duplicated, not sure if this should be removed or if the second one was supposed to be a different hint type.